handler/artifact_upload_handler.py

```diff
@@ -45,24 +48,27 @@ def _check_token(self, token):
         # Reduce the risk of timing analysis attacks.
         return hmac.compare_digest(self.hash_expected, hash_result)
 
-    def _index_channel(self):
+    def _index_channel(self, channel):
         """
-        Index the conda channel at `self.write_path` to ensure the channel
-        remains consistent.
+        Index the conda channel to ensure that it remains consistent.
 
         """
-        cmds = [self.conda_exe, 'index', self.write_path]
+        cmds = [self.conda_exe, 'index', channel]
         subprocess.check_call(cmds)
 
     def post(self):
         token = self.get_argument('token')
+        arch = self.get_argument('arch', default='none')
+        arch = '' if arch == 'none' else arch
         file_data, = self.request.files['artifact']
         filename = file_data['filename']
         body = file_data['body']
         if self._check_token(token):
-            with open(os.path.join(self.write_path, filename), 'wb') as owfh:
+            channel = os.path.join(self.write_path, arch)
```

Comment:
I may be being a little fussy, it might be a good idea to update the explanation of what `write_path` is now that it's pointing to the parent directory of the directory that is getting written to rather than the directory itself.
By which I mean
`write_path` could originally have been: `http://anaconda-channel/linux64`
but now `it is just `http://anaconda-channel`

Reply:
`self.write_path` doesn't ever change. It's always what has been provided from the command line.

The directory that is written to is either `self.write_path` or `self.write_path/arch` (so to speak) if `arch` is provided by the client.

I'll add some comments for clarification.

Reply:
Sorry my previous comment was wrong.

I was trying to say that write_dir says "directory to write artifacts to" but we're not writing artifacts to that directory. We are writing the artifacts to a subdirectory within write_dir.

But as I say this may be a bit fussy

Reply:
Just ignore where I said "By which I mean
write_path could originally have been: http://anaconda-channel/linux64 but now it is just http://anaconda-channel"

Reply:
Yeah, the thing is that it could write to that directory, if the client doesn't provide an `arch`, as `arch` is optional. Kinda depends on what the client provides.

There's no harm in adding some additional comments to clarify :smile:

Reply:
> if the client doesn't provide an arch, as arch is optional.

This artifact uploader is specifically for conda_channels so you should always write to a arch "subchannel". There's no reason to write write to the parent directory. That being said I don't mind it staying optional.

Reply:
I now retract my retraction. 
I believe 
> write_path could originally have been: http://anaconda-channel/linux64 but now it is just http://anaconda-channel

is valid. 
the server we have running was presumably started with write_path set to be `http://anaconda-channel/linux64` but now we need to restart it to be `http://anaconda-channel/`

Reply:
No. I think you need to retract, retract your retraction.

`write_path` is set to a directory, not a URL from the command line. The use case is specifically to support us, and we're using `--write_dir="/this/is/a/channel/directory/linux-64"` at the moment.

But this will change to be `--write_dir="/this/is/a/channel/directory"` and the client will provide an `arch` of either `linux-64` or `noarch`.

Hope that clarifies.

Reply:
My point wasn't about URL/dir (I admit that was a mistake but not relevant), so bar that my original comment was valid.

My point was that `write_dir` used to refer to the location we were writing to, and hence the comment "directory to write artifacts to" being valid as we were setting this to point to the subdirectory of the channel which we were writing to. Now we are pointing to the channel itself which we're not writing to, we're writing to a subdir of that.

Reply:
And they shouldn't be writing to `--write_dir="/this/is/a/channel/directory"` because that is wrong, you should only ever write to a subdir of a channel